dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java

```diff
@@ -1729,6 +2001,40 @@ protected SolrQuery resolveToSolrQuery(Context context, DiscoverQuery discoveryQ
 
         }
 
+        boolean isWorkspace = StringUtils.startsWith(discoveryQuery.getDiscoveryConfigurationName(),
+                DISCOVER_WORKSPACE_CONFIGURATION_NAME);
+        boolean isWorkflow = StringUtils.startsWith(discoveryQuery.getDiscoveryConfigurationName(),
+                DISCOVER_WORKFLOW_CONFIGURATION_NAME);
+        EPerson currentUser = context.getCurrentUser();
+        // Retrieve all the groups the current user is a member of !
+        Set<Group> groups;
+        try {
+            groups = groupService.allMemberGroupsSet(context, currentUser);
+        } catch (SQLException e) {
+            throw new org.dspace.discovery.SearchServiceException(e.getMessage(), e);
+        }
+
+        // extra security check to avoid the possibility that an anonymous user
+        // get access to workspace or workflow
+        if (currentUser == null && (isWorkflow || isWorkspace)) {
+            throw new IllegalStateException("An anonymous user cannot perform a workspace or workflow search");
+        }
+        if (isWorkspace) {
+            // insert filter by submitter
+            solrQuery
+                .addFilterQuery("submitter:(" + currentUser.getID() + ")");
+        } else if (isWorkflow) {
+            // insert filter by controllers
+            StringBuilder controllerQuery = new StringBuilder();
+            controllerQuery.append("taskfor:(e" + currentUser.getID());
+            for (Group group : groups) {
+                controllerQuery.append(" OR g").append(group.getID());
```

Comment:
Initially we have put this code in the SorlServiceResourceRestrictionPlugin but it was noted that the scope was not exactly the same. We agree and for this reason we moved it. We have considered the idea of a specific plugin but at the end we discarded it because this behavior is not something that can be turnoff as doing that will make the SolrServiceImpl inconsinstent so despite I agree that we should make the SolrServiceImpl more modular in this case we cannot reuse the org.dspace.discovery.SolrServiceIndexPlugin infrastructure that should only add "additionalIndex" and not core information.
I found this old ticket https://jira.duraspace.org/browse/DS-3093 but I also that @benbosman should have opened or will open soon a more detailed ticket about some proposed refactoring